src/transformers/models/llama/modeling_llama.py

```diff
@@ -41,6 +41,143 @@
 from .configuration_llama import LlamaConfig
 
 
+class AttentionMaskCache:
```

Comment:
Very cool class! Once porting it to other models, is the idea to have this upstreamed somewhere or would you rather keep it in this class and copy/paste it around?

Reply:
I've now added also logic to do the windowed attention mask here as well because windowed attention is allowed to be both 2D in Flash Attention and 4D (cc @ArthurZucker)

Besides causal attn mask, normal attn mask, windowed attention mask, I don't think there are any other masks that are supported by FA2 so this class should be able to handle all use cases.

Mid-term I think we should move this attention mask cache to a general utils as it should be able to handle all use cases. For now, I'd like to leave it here until we've added FA2 to all important models.